gcc/rust/expand/rust-macro-builtins.cc

```diff
@@ -61,13 +63,115 @@ macro_end_token (AST::DelimTokenTree &invoc_token_tree,
   return last_token_id;
 }
 
+/* Expand and extract an expression from the macro */
+
+static inline AST::ASTFragment
+try_expand_macro_expression (AST::Expr *expr, MacroExpander *expander)
+{
+  rust_assert (expander);
+
+  auto vis = Rust::AttrVisitor (*expander);
+  expr->accept_vis (vis);
+  return expander->take_expanded_fragment (vis);
+}
+
+/* Expand and then extract a string literal from the macro */
+
+static std::unique_ptr<AST::LiteralExpr>
+try_extract_string_literal_from_fragment (const Location &parent_locus,
+					  std::unique_ptr<AST::Expr> node)
```

Comment:
```suggestion
					  std::unique_ptr<AST::Expr>& node)
```

This would enable us to only clone the expression if the expansion is successful.

Then, this requires the following changes...

Reply:
This, unfortunately, does not work.

Changing this to a reference type will now require an `lvalue` for the 2nd argument (`node` here), and then:

https://github.com/Rust-GCC/gccrs/blob/4af2d3228e6933f9c0736c9ac587e77aec8f6186/gcc/rust/expand/rust-macro-builtins.cc#L109-L111

`nodes.take_expression_fragment ()` becomes invalid since it cannot be made into an `lvalue` according to the compiler.

I have to do this to workaround the compiler's complaint, but I am not sure if this is semantically correct:

```cpp
  auto expr = nodes.take_expression_fragment ();
  return try_extract_string_literal_from_fragment (
    input_expr->get_locus (), expr);
```

Reply:
Yes, this is completely valid. I think if it's not too annoying it would be better to use that, if not feel free to disregard :)